### ISSUE TITLE
False negative with nested lambdas

### ISSUE THREAD

**Issue by dwoffinden** [typetools/checker-framework#5541]
Following https://checkerframework.org/manual/#reporting-bugs, 
 * commands (that can be cut-and-pasted into a command shell):
 
 ```
 ~/checker-framework-3.29.0/checker/bin/javac -processor nullness Test.java && java Test
 ```
 
 * inputs:

`Test.java`:
```
// Copyright 2023 Google LLC.
// SPDX-License-Identifier: Apache-2.0

import java.util.function.Function;
import java.util.function.Supplier;
import org.checkerframework.checker.nullness.qual.Nullable;

final class Test {
  static boolean foo(PromiseRunner runner) {
    return runner.run(
        () ->
            immediatePromise(true)
                .thenChain(
                    b -> {
                      if (b) {
                        return immediatePromise(null);
                      }
                      return immediatePromise(false);
                    }));
  }
  private static <R extends @Nullable Object> Promise<R> immediatePromise(R value) {
    return new Promise<>(value);
  }
  private static final class PromiseRunner {
    <T> T run(Supplier<? extends Promise<? extends T>> supplier) {
      return supplier.get().value;
    }
  }
  private static final class Promise<T extends @Nullable Object> {
    final T value;
    private Promise(T value) {
      this.value = value;
    }
    <T2 extends @Nullable Object> Promise<T2> thenChain(
        Function<? super T, Promise<T2>> dataTransform) {
      return dataTransform.apply(value);
    }
  }
  public static void main(String[] args) {
    boolean unused = foo(new PromiseRunner());
  }
}
```

 * outputs:
 
 Code fails with a NPE:
 ```
 Exception in thread "main" java.lang.NullPointerException: Cannot invoke "java.lang.Boolean.booleanValue()" because the return value of "Test$PromiseRunner.run(java.util.function.Supplier)" is null
	at Test.foo(Test.java:7)
	at Test.main(Test.java:43)
```

 * expectation:
 
I would've expected the checker to fail -- the return type of `immediatePromise(null)` is clearly `Promise<@Nullable Boolean>` and so `run()` can return `null`.

**Comment by smillst** [comment]
An error is now correctly issued for this test case.
